dkulib/parallelizer/parallelizer.py

```diff
@@ -0,0 +1,278 @@
+    output_df = pd.DataFrame.from_records(record_list).astype(output_schema).reindex(columns=output_column_list)
+    return output_df
+
+
+def parallelizer(
+    input_df: pd.DataFrame,
+    function: Callable,
+    exceptions: Union[Exception, Tuple[Exception]],
+    column_prefix: AnyStr,
+    parallel_workers: int = DEFAULT_PARALLEL_WORKERS,
+    batch_support: bool = DEFAULT_BATCH_SUPPORT,
+    batch_size: int = DEFAULT_BATCH_SIZE,
+    error_handling: ErrorHandling = ErrorHandling.LOG,
+    verbose: bool = DEFAULT_VERBOSE,
+    **function_kwargs,
+) -> pd.DataFrame:
+    """Apply a function to a pandas.DataFrame with parallelization, batching, error handling and progress tracking
+
+    The DataFrame is iterated on and passed to the function as dictionaries, row-by-row or by batches of rows.
+    This iterative process is accelerated by the use of concurrent threads and is tracked with a progress bar.
+    Errors are catched if they match the `exceptions` parameter and automatically logged.
+    Once the whole DataFrame has been iterated on, results and errors are added as additional columns.
+
+    Args:
+        input_df: Input dataframe which will be iterated on
+        function: Function taking a dict as input and returning a dict
+            If `function_support_batch` then the function works on list of dict
+            For instance, a function to call an API or do some enrichment
+        exceptions: Tuple of Exception classes to catch
+        column_prefix: Column prefix to add to the output columns for the `function` responses and errors
+        parallel_workers: Number of concurrent threads
+        batch_support: If True, send batches of row to the `function`
+            Else (default) send rows as dict to the function
+        batch_size: Number of rows to include in each batch
+            Taken into account if `batch_support` is True
+        error_handling: If ErrorHandling.LOG (default), log the error message as a warning
+            and return the row with error keys.
+            Else fail is there is any error.
+        verbose: If True, log additional information on errors
+            Else (default) log the error message and the error type
+        **function_kwargs: Arbitrary keyword arguments passed to the `function`
+
+    Returns:
+        Input dataframe with additional columns:
+        - response from the `function`
+        - error message if any
+        - error type if any
+
+    """
+    df_iterator = (index_series_pair[1].to_dict() for index_series_pair in input_df.iterrows())
+    len_iterator = len(input_df.index)
+    start = perf_counter()
+    if batch_support:
+        logging.info(
+            f"Applying function {function.__name__} in parallel to {len_iterator} row(s)"
+            + f" using batch size of {batch_size}..."
+        )
+        df_iterator = chunked(df_iterator, batch_size)
+        len_iterator = math.ceil(len_iterator / batch_size)
```

Comment:
I think it is clearer to rename rather than re assigne

Reply:
Agreed, done in  b7157df